File: test/cpu/neg.cc
#define BOOST_TEST_DYN_LINK
#include "68040.hpp"

#include "test.hpp"
#include <boost/test/data/monomorphic.hpp>
#include <boost/test/data/test_case.hpp>
#include <boost/test/unit_test.hpp>
namespace bdata = boost::unit_test::data;
struct F_NEG {
    F_NEG() {
        // NEG.B %D2
        TEST::SET_W(0, 0042000 | 2);
        TEST::SET_W(2, TEST_BREAK);

        // NEG.W %D2
        TEST::SET_W(4, 0042100 | 2);
        TEST::SET_W(6, TEST_BREAK);

        // NEG.L %D2
        TEST::SET_W(8, 0042200 | 2);
        TEST::SET_W(10, TEST_BREAK);

        jit_compile(0, 12);
    }};
BOOST_FIXTURE_TEST_SUITE(NEG, Prepare, *boost::unit_test::fixture<F_NEG>())



BOOST_AUTO_TEST_SUITE(Byte)

BOOST_AUTO_TEST_CASE(value) {
    cpu.D[2] = -2;
    run_test(0);
    BOOST_TEST((cpu.D[2] & 0xff) == 2);
    BOOST_TEST(!cpu.V);
    BOOST_TEST(cpu.C);
    BOOST_TEST(cpu.X);
    BOOST_TEST(!cpu.N);
    BOOST_TEST(!cpu.Z);
}

BOOST_AUTO_TEST_CASE(V) {
    cpu.D[2] = 0x80;
    run_test(0);
    BOOST_TEST(cpu.V);
}

BOOST_AUTO_TEST_CASE(N) {
    cpu.D[2] = 3;
    run_test(0);
    BOOST_TEST(cpu.N);
}

BOOST_AUTO_TEST_CASE(Z) {
    cpu.D[2] = 0;
    run_test(0);
    BOOST_TEST(cpu.Z);
    BOOST_TEST(!cpu.C);
    BOOST_TEST(!cpu.X);
}

BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE(Word)


BOOST_AUTO_TEST_CASE(value) {
    cpu.D[2] = -200;
    run_test(4);
    BOOST_TEST((cpu.D[2] & 0xffff) == 200);
    BOOST_TEST(!cpu.V);
    BOOST_TEST(cpu.C);
    BOOST_TEST(cpu.X);
    BOOST_TEST(!cpu.N);
    BOOST_TEST(!cpu.Z);
}

BOOST_AUTO_TEST_CASE(V) {
    cpu.D[2] = 0x8000;
    run_test(4);
    BOOST_TEST(cpu.V);
}

BOOST_AUTO_TEST_CASE(N) {
    cpu.D[2] = 100;
    run_test(4);
    BOOST_TEST(cpu.N);
}

BOOST_AUTO_TEST_CASE(Z) {
    cpu.D[2] = 0;
    run_test(4);
    BOOST_TEST(cpu.Z);
    BOOST_TEST(!cpu.C);
    BOOST_TEST(!cpu.X);
}

BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE(Long)



BOOST_AUTO_TEST_CASE(value) {
    cpu.D[2] = -200000;
    run_test(8);
    BOOST_TEST(cpu.D[2] == 200000);
    BOOST_TEST(!cpu.V);
    BOOST_TEST(cpu.C);
    BOOST_TEST(cpu.X);
    BOOST_TEST(!cpu.N);
    BOOST_TEST(!cpu.Z);
}


BOOST_AUTO_TEST_CASE(V) {
    cpu.D[2] = 0x80000000;
    run_test(8);
    BOOST_TEST(cpu.V);
}

BOOST_AUTO_TEST_CASE(N) {
    cpu.D[2] = 10000;
    run_test(8);
    BOOST_TEST(cpu.N);
}

BOOST_AUTO_TEST_CASE(Z) {
    cpu.D[2] = 0;
    run_test(8);
    BOOST_TEST(cpu.Z);
    BOOST_TEST(!cpu.C);
    BOOST_TEST(!cpu.X);
}

BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE_END()

File: test/lua/cpu/neg.lua
-- NEG.B %D2
MEM[0].W = OCT"042000" | 2
MEM[2].W = TEST.BREAK

-- NEG.W %D3
MEM[4].W = OCT"042100" | 3
MEM[6].W = TEST.BREAK

-- NEG.L %D4
MEM[8].W = OCT"042200" | 4
MEM[10].W = TEST.BREAK

JIT_COMPILE(0, 12)

local run = {}

run.B = {
    value = function()
        CPU.D[3].C = -2
        TEST.RUN(0)
        TEST.CHECK("CPU.D[3].C", 2)
    end,
    v = farray(TF, function(b)
        CPU.D[3].C = cond(b, -128, -127)
        TEST.RUN(0)
        TEST.CHECK("CPU.CC.V", b)
    end),    
    n = farray(TF, function(b)
        CPU.D[3].C = cond(b, 5, -5)
        TEST.RUN(0)
        TEST.CHECK("CPU.CC.N", b)
    end),   
    zcx = farray(TF, function(b)
        CPU.D[3].C = cond(b, 0, 1)
        TEST.RUN(0)
        TEST.CHECK("CPU.CC.Z", b)
        TEST.CHECK("CPU.CC.C", not b)
        TEST.CHECK("CPU.CC.X", not b)
    end),    
}

run.W = {
    value = function()
        CPU.D[4].H = -2000
        TEST.RUN(4)
        TEST.CHECK("CPU.D[4].H", 2000)
   end,
   v = farray(TF, function(b)
        CPU.D[4].H = cond(b, 0x8000, 0x8001)
        TEST.RUN(4)
        TEST.CHECK("CPU.CC.V", b)
    end),    
    n = farray(TF, function(b)
        CPU.D[4].H = cond(b, 5000, -5000)
        TEST.RUN(4)
        TEST.CHECK("CPU.CC.N", b)
    end),   
    zcx = farray(TF, function(b)
        CPU.D[4].H = cond(b, 0, 1)
        TEST.RUN(4)
        TEST.CHECK("CPU.CC.Z", b)
        TEST.CHECK("CPU.CC.C", not b)
        TEST.CHECK("CPU.CC.X", not b)
    end),    
}

run.L = {
    value = function()
        CPU.D[5].I = -2000000
        TEST.RUN(8)
        TEST.CHECK("CPU.D[5].I", 2000000)
    end,

    v = farray(TF, function(b)
        CPU.D[5].I = cond(b, 0x80000000, 0x80000001)
        TEST.RUN(8)
        TEST.CHECK("CPU.CC.V", b)
    end),    
    n = farray(TF, function(b)
        CPU.D[5].I = cond(b, 5000000, -5000000)
        TEST.RUN(8)
        TEST.CHECK("CPU.CC.N", b)
    end),   
    zcx = farray(TF, function(b)
        CPU.D[5].I = cond(b, 0, 1)
        TEST.RUN(8)
        TEST.CHECK("CPU.CC.Z", b)
        TEST.CHECK("CPU.CC.C", not b)
        TEST.CHECK("CPU.CC.X", not b)
    end),    
}
return run